[package] track storages across lifetime of PackageExporter

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59735**
* #59734

General changes:
1.  Changed `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs. Serialization needs to recognize storages by the `c10::Storage` object reference and provide for a deterministic naming scheme. Deserialization needs to be able to map a filename to a loaded `c10::Storage`.
2. Added determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter created in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

torch.package specific changes:

3. Fixed ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

Commit:
[package] track storages across lifetime of PackageExporter

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"




[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"




[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


Fixes potential ABA storage identity problem during serialization. Solves by keeping reference to all storages seen during a PackageExporter's lifetime, using same mechanics as the solution for Mobile.

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` using same method employed in `tensor_cdata_naming_scheme` (keeps reference of storage through lifetime of `PackageExporter` to prevent reuse of memory address). 
2. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` using same method employed in `tensor_cdata_naming_scheme` (keeps reference of storage through lifetime of `PackageExporter` to prevent reuse of memory address). 
2. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` using same method employed in `tensor_cdata_naming_scheme` (keeps reference of storage through lifetime of `PackageExporter` to prevent reuse of memory address). 
2. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue. 
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`)
3. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue. 
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`)
3. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue. 
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`). Now the storage id is based off of a counter instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage
3. Additionally uses presence of a storage in the `StorageContext` instance as marker for if a storage has been serialized or not, removing the need to scan the `PythonStreamWriter` for presence of the storage's serialization file

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `StorageContext` embedded in the `ScriptModuleDeserializer` instance).
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`). Now the storage id is based off of a counter sourced in `StorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage
3. Removes need to scan `PythonStreamWriter` for presence of storage files by using the presence of a storage in the `StorageContext` instance as a marker that the storage has been serialized already.

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `StorageContext` embedded in the `ScriptModuleDeserializer` instance).
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`). Now the storage id is based off of a counter sourced in `StorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage
3. Removes need to scan `PythonStreamWriter` for presence of storage files by using the presence of a storage in the `StorageContext` instance as a marker that the storage has been serialized already.

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

suo commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -994,7 +995,24 @@ void initJitScriptBindings(PyObject* module) {
+            return m.storage_context().getStorageID(name);
+          })
+      .def(
+          "track_storage",
```
I think this should be `add_storage`, to better map onto the C++ API.

suo commented on torch/csrc/jit/serialization/import.h:
```diff
@@ -17,27 +17,46 @@ class ReadAdapterInterface;
-    storage_map_.insert({name, storage});
+  explicit StorageContext() : unique_id(0) {}
+
+  uint64_t addStorage(const std::string& name, c10::Storage storage) {
```
I think the API for this can be made simpler by removing the concept of names. 
- addStorage(storage) -> id
- hasStorage(id) -> bool
- getId(storage) -> id
- getStorage(id) -> storage

We don't really need names, right? Since we plan to serialize the id anyway.

Lilyjjo replied:
The names help solve a weird situation the deserialization process. Basically during unpicking we have the storage’s ‘name’ returned from the storage’s persid tuple and that’s it. In the `StorageContext` we need a way for both the TS unpickler and torch.package’s load_tensor function to be able to map this outside-the-storage-context key to the same storage, and that’s not doable with the unique_id if the unique_id is being generated from the `StorageContext` itself. The weirdness of this StorageContext is really coming from us trying to use it for both serialization and deserialization which have different needs (serialization needing the unique_id and the deseralziation needing to map arbitrary names to storages)

suo commented on torch/package/package_exporter.py:
```diff
@@ -724,18 +722,19 @@ def _persistent_id(self, obj):
-            return ("storage", storage_type, obj_key, location, obj.size())
+
+            # serialize storage if not already written
+            if not self.script_module_serializer.has_storage(obj_key):
```
We don't have to preserve the cdata naming scheme here I think. We ought to be consistent with TS and use `StorageContext` to assign names to tensors during serialization.

Lilyjjo replied:
I can write comments on this code, it is confusing. The cdata isn't being used for the storage name, it's just being used to identify the storage itself across cpp/python during serialization. This code is consistent with the TS code of using the `storage_id` as the filename. (Is that what your comment was asking about?)

suo commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -475,11 +477,19 @@ void ScriptModuleSerializer::writeArchive(
-          tensor_names.push_back(string_id + ".storage");
-          storage_context_.addStorage(string_id, tensor.storage());
+          uint64_t id = 0;
+          if (storage_context_.hasStorage(cdata_str)) {
```
If we use StorageContext to assign names in `torch.package`, we don't need this branch.

suo commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -463,6 +463,8 @@ void ScriptModuleSerializer::writeArchive(
   std::vector<c10::ClassTypePtr> memoizedClassTypes;
   std::vector<std::string> tensor_names;
+  // tensors which need their storages written in tensor_cdata_naming_scheme
+  std::unordered_set<std::string> tensors_to_serialize;
```
nit: since the only time we are using this is to check that something is *not* in it, can we invert this and make it `already_serialized`.

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `StorageContext` embedded in the `ScriptModuleDeserializer` instance).
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`). Now the storage id is based off of a counter sourced in `StorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage
3. Removes need to scan `PythonStreamWriter` for presence of storage files by using the presence of a storage in the `StorageContext` instance as a marker that the storage has been serialized already.

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `StorageContext` embedded in the `ScriptModuleDeserializer` instance).
2. Adds determinism to naming scheme of serialized storages in export code paths which utilize `tensor_cdata_naming_scheme`(introduced 2nd mapping in `StorageContext`, now maps `storage cdata ptr` -> `unique id`, `unique id` -> `c10::Storage`). Now the storage id is based off of a counter sourced in `StorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage
3. Removes need to scan `PythonStreamWriter` for presence of storage files by using the presence of a storage in the `StorageContext` instance as a marker that the storage has been serialized already.

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Changes `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs
2. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).
3. Adds determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter sourced in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

suo commented on torch/csrc/jit/serialization/storage_context.h:
```diff
@@ -0,0 +1,72 @@
+  ~DeserializationStorageContext() = default;
+
+ private:
+  std::map<std::string, c10::Storage> name_storage_map_;
```
`unordered_map` is slightly better here. It has better insertion and lookup performance, and we do not care about the ordering.

suo commented on torch/csrc/jit/serialization/storage_context.h:
```diff
@@ -0,0 +1,72 @@
+  };
+
+  uint64_t unique_id;
+  std::map<c10::Storage, uint64_t, StorageSerializationComparator>
```
small comments:
1. `std::unordered_map` is slightly better here, since it has better lookup/retrieval performance and we don't care about ordering.
2. We don't need `unique_id`. Since we never delete storage from the serialization context, we can just do `storage_id_map_[storage] = storage_id_map_.size()`.

suo commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -987,14 +988,31 @@ void initJitScriptBindings(PyObject* module) {
           "write_files",
           &ScriptModuleSerializer::writeFiles,
-          py::arg("code_dir") = ".data/ts_code/code/");
+          py::arg("code_dir") = ".data/ts_code/code/")
```
Let's create a separate binding for `SerializationStorageContext`, for two reasons:
1. It's more readable/understandable for there to be a 1:1 mapping of pybind classes to cpp classes, with the same API exposed.
2. It currently gives the false impression that we only need `SerializationStorageContext` in the context of ScriptModule serialization.

suo commented on torch/csrc/jit/serialization/storage_context.h:
```diff
@@ -0,0 +1,72 @@
+ public:
+  explicit SerializationStorageContext() : unique_id(0) {}
+
+  uint64_t addStorage(c10::Storage storage) {
```
Seems like the API would be cleaner with just `getOrAddStorage`, which either retrieves an id from the table or creates a new one. Some benefits:
1. It is no longer possible to use the API incorrectly (e.g. those `TORCH_INTERNAL_ASSERT`s you have now will no longer need to exist). It's nice to make APIs that are impossible to use incorrectly.
2. It simplifies user code (you can get rid of the if-else branch in export_module.cpp:477)

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Changes `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs
2. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).
3. Adds determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter sourced in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Changes `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs
2. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).
3. Adds determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter sourced in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


1. Changes `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs
2. Fixes ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).
3. Adds determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter created in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Lilyjjo commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -472,12 +474,19 @@ void ScriptModuleSerializer::writeArchive(
+            serialized_tensors.insert(tensor_name);
+          }
+          tensor_names.push_back(tensor_name);
         } else {
```
This is the part of the change that affects the mobile code

cccclai replied:
It should also change the model generated from torch.package? It looks reasonable to me. I've run through a unit test and check one simple model. It seems ok. I may need to run some stress tests for production models. We don't have an automate way at this moment :/

Lilyjjo replied:
It changes the torch.package archive as well!

raziel replied:
Just want to double check @cccclai you ran those tests and/or are ok with the change.

cccclai replied:
Yeah I ran the test for existing prod models and they look good from mobile side

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -210,8 +210,8 @@ void writeArchiveV5(
-    bool tensor_cdata_naming_scheme,
-    StorageContext& storage_context) {
+    bool use_storage_context,
+    SerializationStorageContext& storage_context) {
```
Thanks for catching it!

suo reviewed (approved):
lgtm!

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


General changes:
1.  Changed `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs. Serialization needs to recognize storages by the `c10::Storage` object reference and provide for a deterministic naming scheme. Deserialization needs to be able to map a filename to a loaded `c10::Storage`.
2. Added determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter created in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

torch.package specific changes:

3. Fixed ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


General changes:
1.  Changed `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs. Serialization needs to recognize storages by the `c10::Storage` object reference and provide for a deterministic naming scheme. Deserialization needs to be able to map a filename to a loaded `c10::Storage`.
2. Added determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter created in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

torch.package specific changes:

3. Fixed ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

cccclai reviewed (approved):
Ran through existing prod models by resave/load/run. Looks good from mobile side.

iseeyuan commented:
> Ran through existing prod models by resave/load/run. Looks good from mobile side.

Thanks @cccclai for verifying this PR! We should put prod model serialization tests in CI, so that you don't have to run the verificatoin manually.

Commit:
Update on "[package] track storages across lifetime of PackageExporter"


General changes:
1.  Changed `StorageContext` into `SerializationStorageContext` and `DeserializationStorageContext` since APIs have different needs. Serialization needs to recognize storages by the `c10::Storage` object reference and provide for a deterministic naming scheme. Deserialization needs to be able to map a filename to a loaded `c10::Storage`.
2. Added determinism to naming scheme of serialized storages in export code paths by changing `tensor_cdata_naming_scheme` to `use_storage_context`. Now the storage id is based off of a counter created in the `SerializationStorageContext` instead of the memory address.
-     .data/90987234290.storage --> .data/0.storage

torch.package specific changes:

3. Fixed ABA storage identity problem during serialization for `torch.package` by keeping reference of serialized storages through lifetime of `PackageExporter` to prevent reuse of memory address. Achieved by extending logic used in solution to mobile's same issue (i.e. using the `SerializationStorageContext` embedded in the `ScriptModuleSerializer` instance).

Differential Revision: [D29075276](https://our.internmc.facebook.com/intern/diff/D29075276)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29075276).
